### PR TITLE
#532 Declarative fix of AvgOf() bug on Double edge values

### LINKED ISSUE #532: AvgOf produces overflow for large values
```
new AvgOf(
          new ListOf<>(Double.MAX_VALUE, Double.MAX_VALUE).toArray(new Double[2])
).doubleValue();
``` 
returns an `Infinity` but should be `Double.MAX_VALUE`

### PR DESCRIPTION
**Pull request by rexim**
#532 is mine now: https://github.com/yegor256/cactoos/issues/532#issuecomment-360486873 Based on #543 

Close #532 close #543

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#581](https://github.com/yegor256/cactoos/issues/581) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#581](https://github.com/yegor256/cactoos/issues/581) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 1]
```diff
@@ -71,207 +74,51 @@
...
-                total += 1.0d;
-            }
-            if (total == 0.0d) {
-                total = 1.0d;
-            }
-            return sum / total;
-        });
+        this(
+            new Mapped<>(
+                (number) -> () -> number,
```
Comment (neonailol) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: unnecessary parenthises, does not correspond with style used in library

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 2]
```diff
@@ -289,59 +136,25 @@ public AvgOf(final Scalar<Number>... src) {
...
-            }
-            if (total == 0.0d) {
-                total = 1.0d;
-            }
-            return sum / total;
-        });
+        super(
+            new Ternary<>(
+                new LengthOf(src).longValue(),
+                (len) -> len > 0,
```
Comment (neonailol) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: unnecessary parenthises, does not correspond with style used in library

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 3]
```diff
@@ -289,59 +136,25 @@ public AvgOf(final Scalar<Number>... src) {
...
+                (len) -> new Reduced<>(
```
Comment (neonailol) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: unnecessary parenthises, does not correspond with style used in library

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 4]
```diff
@@ -289,59 +136,25 @@ public AvgOf(final Scalar<Number>... src) {
...
+                    BigDecimal.ZERO,
+                    BigDecimal::add,
+                    new Mapped<>(
+                        (number) -> BigDecimal.valueOf(
```
Comment (neonailol) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: unnecessary parenthises, does not correspond with style used in library

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 5]
```diff
@@ -289,59 +136,25 @@ public AvgOf(final Scalar<Number>... src) {
...
+                            number.value().doubleValue()
+                        ),
+                        src
+                    )
+                ).value().divide(
+                    BigDecimal.valueOf(len),
+                    MathContext.DECIMAL128
+                ).doubleValue(),
+                (len) -> 0.0
```
Comment (neonailol) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: unnecessary parenthises, does not correspond with style used in library

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 6]
```diff
@@ -51,125 +51,308 @@ public void withEmptyCollection() {
...
+        MatcherAssert.assertThat(
+            new AvgOf(
+                () -> 1.0d, () -> 2.0d, () -> 10.0d
+            ).intValue(),
+            Matchers.equalTo(4)
+        );
+    }
+
+    @Test
+    public void withIterableOfScalars() {
```
Comment (neonailol) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: in this test type of `new IterableOf<Scalar<Number>>(() -> 1L, () -> 2L, () -> 10L)` can be inferred, so params are not neccecary

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 7]
```diff
@@ -178,4 +361,16 @@ public void withScalars() {
...
     }
 
+    @Test
+    public void withCompositionOfScalars() {
+        MatcherAssert.assertThat(
+            new AvgOf(
+                () -> new MinOf(1.0d, 2.0d),
+                () -> new MaxOf(2.0d, 4.0d),
+                () -> new SumOf(1.0d, 2.0d, 2.0d),
+                new Ternary<>(true, 5.0, 1.0d)
```
Comment (neonailol) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: why just `5.0` without `d` or why `d` is need here at all?

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 8]
```diff
@@ -51,125 +51,308 @@ public void withEmptyCollection() {
...
+            ).intValue(),
+            Matchers.equalTo(4)
+        );
+    }
+
+    @Test
+    public void withFloatScalarsIntValue() {
+        MatcherAssert.assertThat(
+            new AvgOf(
+                () -> 1f, () -> 2f, () -> 10f
```
Comment (neonailol) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: please format floats as `1.0f`

File: src/test/java/org/cactoos/scalar/AvgOfTest.java (hunk 6 above)
Comment (rexim) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: Without it the compiler complains:

```
[WARNING] yegor256/cactoos/src/test/java/org/cactoos/scalar/AvgOfTest.java:[358,17] unchecked generic array creation for varargs parameter of type org.cactoos.Scalar<java.lang.Number>[]
```

And we have `-Werror` enabled.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java (hunk 7 above)
Comment (rexim) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: @neonailol typo. They were in the original code I inherited. I kept using the for the consistency. I can remove them if it does not correspond with the style of the library. :)

**Comment by rexim** [comment]
@neonailol fixed your remarks. Please take a look again.

**Comment by neonailol** [comment]
Thank you, looks good to me

**Comment by neonailol** [comment]
@yegor256 this pull request can be merged into master

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol), total is [+180](https://www.0crat.com/u/neonailol)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+8755](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#581](https://github.com/yegor256/cactoos/issues/581) is now out of scope

**Pull request merged**